frame/dapps-staking/src/pallet/mod.rs

```diff
@@ -659,6 +665,90 @@ pub mod pallet {
             Ok(().into())
         }
 
+        /// Transfer nomination from one contract to another.
+        ///
+        /// Same rules as for `bond_and_stake` and `unbond_and_unstake` apply.
+        /// Minor difference is that there is no unbonding period so this call won't
+        /// check whether max number of unbonding chunks is exceeded.
+        ///
+        #[pallet::weight(T::WeightInfo::nomination_transfer())]
+        pub fn nomination_transfer(
+            origin: OriginFor<T>,
+            origin_contract_id: T::SmartContract,
+            #[pallet::compact] value: BalanceOf<T>,
```

Comment:
Let’s permit all balance transfers only, it will make code much more simple and safe.

Reply:
We can do that, it's no problem to implement it.
However, we won't reduce complexity of the code.
We will only reduce what the user can do and possibly, I'll have to redo UTs later if users request this feature.

The `bond_and_stake` and `unbond_and_unstake` logic is completely reused here.
The only _new_ part of the code are the checks specific for the nomination transfer.

So if we decide only to allow full transfer, this extrinsic call will change to this:
```Rust
pub fn nomination_transfer(
            origin: OriginFor<T>,
            origin_contract_id: T::SmartContract,
            target_contract_id: T::SmartContract,
        ) -> DispatchResultWithPostInfo {
        // same code as before...
        let mut origin_staker_info = Self::staker_info(&staker, &origin_contract_id);
        let value = origin_staker_info.latest_staked_value();
        // same code as before
}
```